pkg/logql/evaluator.go

```diff
@@ -898,3 +949,31 @@ func literalStepEvaluator(
 		eval.Error,
 	)
 }
+
+// absentLabels guesses labels based on labels matchers of a SampleExpr
+// This is to replace missing timeseries during absent_over_time aggregation.
+func absentLabels(expr SampleExpr) labels.Labels {
+	m := labels.Labels{}
+
+	lm := expr.Selector().Matchers()
+	if len(lm) == 0 {
+		return m
+	}
+
+	empty := []string{}
+	for _, ma := range lm {
+		if ma.Name == labels.MetricName {
+			continue
+		}
+		if ma.Type == labels.MatchEqual && !m.Has(ma.Name) {
+			m = labels.NewBuilder(m).Set(ma.Name, ma.Value).Labels()
+		} else {
+			empty = append(empty, ma.Name)
+		}
+	}
+
+	for _, v := range empty {
+		m = labels.NewBuilder(m).Del(v).Labels()
```

Comment:
I think it's a bit unusual to use a new builder for each label iteration, but I don't think this is a bottleneck either.

Reply:
This is how Prometheus does it.